### ISSUE TITLE
 Bug: [prefer-readonly-parameter-types] `isTypeReadonly` stack overflow

### ISSUE THREAD

**Issue by sviat9440** [typescript-eslint/typescript-eslint#5875]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=JYOwLgpgTgZghgYwgAgCoQM5mQbwLABQyyUEcAJgPYgA2AnsgNoDWEdAXMllKAOYC6ndFgCMAbkIBfQoVCRYiFMLAjchYqQrV6TVhy5geIAUMxgJBaQSA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0Wls4CGAEwD2TeIXRRe0EdEjgwAXxBKgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
interface Test {
  readonly [key: string]: Test1;
}

interface Test1 {
  readonly [key: string]: Test;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {},
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {}
}
```

### Expected Result

`isTypeReadonly` for type `Test` should return `true`

### Actual Result

Error:
```
RangeError: Maximum call stack size exceeded
    at getReducedApparentType (C:\Users\User\IdeaProjects\typescript-eslint\node_modules\typescript\lib\typescript.js:59926:40)
    at Object.getPropertiesOfType (C:\Users\User\IdeaProjects\typescript-eslint\node_modules\typescript\lib\typescript.js:59550:20)
    at TypeObject.Object.<anonymous>.TypeObject.getProperties (C:\Users\User\IdeaProjects\typescript-eslint\node_modules\typescript\lib\typescript.js:166905:33)
    at getProperties (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:131:27)
    at isTypeReadonlyObject (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:301:28)
    at isTypeReadonlyRecurser (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:120:14)
    at checkIndexSignature (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:203:36)
    at isTypeReadonlyObject (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:301:28)
    at isTypeReadonlyRecurser (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:120:14)
    at checkIndexSignature (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:203:36)
    at isTypeReadonlyObject (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:301:28)
    at isTypeReadonlyRecurser (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:120:14)
    at checkIndexSignature (C:\Users\User\IdeaProjects\typescript-eslint\packages\type-utils\src\isTypeReadonly.ts:203:36)
```

### Additional Info

Related to #4476.

### Versions

Not relevant

**Comment by JoshuaKGoldberg** [comment]
❤️‍🔥  thanks for filing!
